sources/packages/api/src/services/application/application.service.ts

```diff
@@ -638,4 +646,129 @@ export class ApplicationService extends RecordDataModelService<Application> {
       });
     return query.getOne();
   }
+
+  /**
+   * Creates a new Student Application overriding the current one
+   * in order to rollback the process and start the assessment
+   * all over again.
+   * @param locationId location id executing the COE rollback.
+   * @param applicationId application to be rolled back.
+   * @returns the overridden and the newly created application.
+   */
+  async overrideApplicationForCOE(
+    locationId: number,
+    applicationId: number,
+  ): Promise<ApplicationOverriddenResult> {
+    const appToOverride = await this.repo.findOne(
+      {
+        id: applicationId,
+        location: { id: locationId },
+      },
+      { relations: ["studentFiles"] },
+    );
+
+    if (!appToOverride) {
+      throw new CustomNamedError(
+        "Student Application not found or the location does not have access to it",
+        APPLICATION_NOT_FOUND,
+      );
+    }
+
+    if (
+      appToOverride.applicationStatus !== ApplicationStatus.enrollment ||
+      appToOverride.coeStatus !== COEStatus.required
+    ) {
+      throw new CustomNamedError(
+        `Student Application is not in the expected status. The application must be in application status '${ApplicationStatus.enrollment}' and COE status '${COEStatus.required}' to be override.`,
+        INVALID_OPERATION_IN_THE_CURRENT_STATUS,
+      );
+    }
+
+    // Change status of the current application being overwritten.
+    appToOverride.applicationStatus = ApplicationStatus.overwritten;
+    appToOverride.applicationStatusUpdatedOn = getUTCNow();
+
+    // Create a new application based in the previous one.
+    const newApplication = new Application();
+    newApplication.data = appToOverride.data;
+    newApplication.applicationNumber = appToOverride.applicationNumber;
+    newApplication.student = {
```

Comment:
I think we can directly assign the object,  `newApplication.student = appToOverride.student`. can you check

Reply:
Here I took a different approach introducing the @RelationId to the relationships.
The main query does not contain the joins and the relations objects are not being loaded, in this way, we have direct access to the foreign keys that need to be associated with the cloned record. Please let me know if it makes sense.

![image](https://user-images.githubusercontent.com/61259237/130270362-2175217c-7688-4c2b-a310-7211bd5a032a.png)

Reply:
I see what you are doing here. with Cascade set to true on One-to-One, if we don't supply value for "student" instead just populating studentId, we assume studentId will override and take care "student: Student" doesn't become null.

And what happens if there we populate "student: Student" and does not supply studentId? I hope in either of scenario, it should not delete/set null to foreign key.

I asked this because I have less hands-on in typeorm. And it is slightly different from other orm frameworks which i previously worked.

Reply:
All good @dheepak-aot 😃 
So, the @RelationId works in only one direction as described [here](https://github.com/typeorm/typeorm/blob/master/docs/decorator-reference.md#relationid)
"_Relation id is used only for representation. The underlying relation is not added/removed/changed when chaining the value._"

It is just a shortcut to avoid adding a join to the query when we just need the id of the foreign key. We can use it to read the values but, to create relationships while saving the record we still need to do it in the usual way using the relationship object, student in this case.

@guru-aot this is what we talked about in the last sprint. I believe that we have good advantages doing it is this way and not adding the join just to get the ids 😉